bootstore/src/schemes/v0/fsm.rs

```diff
@@ -0,0 +1,820 @@
+            }
+            State::InitialMember { pkg, .. } => {
+                if rack_uuid != pkg.rack_uuid {
+                    MsgError::RackUuidMismatch {
+                        expected: pkg.rack_uuid,
+                        got: rack_uuid,
+                    }
+                    .into()
+                } else {
+                    ResponseType::Share(Share(pkg.share.clone()))
+                }
+            }
+        };
+
+        self.push_response(from, request_id, response);
+    }
+
+    // Handle a `RequestType::Learn` from a peer
+    fn on_learn(&mut self, now: Instant, from: Baseboard, request_id: Uuid) {
+        let err = match &self.state {
+            State::Uninitialized => Some(MsgError::NotInitialized),
+            State::Learning => Some(MsgError::StillLearning),
+            State::Learned { .. } => Some(MsgError::CannotSpareAShare),
+            State::InitialMember { pkg, .. } => {
+                self.request_manager.new_learn_received_req(
+                    request_id,
+                    now,
+                    pkg.rack_uuid,
+                    pkg.threshold,
+                    from.clone(),
+                    &self.connected_peers,
+                );
+                // We need to gather shares and reconstruct the rack secret.
+                // Therefore we don't send a response immediately.
+                None
+            }
+        };
+        if let Some(err) = err {
+            self.push_response(from, request_id, err.into());
+        }
+    }
+
+    // Handle a `Response` from a peer
+    fn handle_response(
+        &mut self,
+        from: Baseboard,
+        rsp: Response,
+    ) -> Result<Option<ApiOutput>, ApiError> {
+        match rsp.type_ {
+            ResponseType::InitAck => self.on_init_ack(from, rsp.request_id),
+            ResponseType::Share(share) => {
+                self.on_share(from, rsp.request_id, share)
+            }
+            ResponseType::Pkg(pkg) => self.on_pkg(from, rsp.request_id, pkg),
+            ResponseType::Error(error) => {
+                Err(ApiError::ErrorResponseReceived {
+                    from,
+                    state: self.state.name(),
+                    request_id: rsp.request_id,
+                    error,
+                })
+            }
+        }
+    }
+
+    // Handle a `ResposneType::InitAck` from a peer
+    fn on_init_ack(
+        &mut self,
+        from: Baseboard,
+        request_id: Uuid,
+    ) -> Result<Option<ApiOutput>, ApiError> {
+        match self.request_manager.on_init_ack(from.clone(), request_id) {
+            Some(true) => Ok(Some(ApiOutput::RackInitComplete)),
+            Some(false) => Ok(None),
+            None => Err(ApiError::UnexpectedResponse {
+                from,
+                state: self.state.name(),
+                request_id,
+                msg: "InitAck",
+            }),
+        }
+    }
+
+    // Handle a `ResponseType::Pkg` from a peer
+    fn on_pkg(
```

Comment:
Vague feelings: `share` vs `pkg` aren't obvious names for "package acquired during initial bringup" versus "package learned by a post-bringup request".

Maybe use `ResponseType::InitialShare` and `ResponseType::LearnPkg` to be clear (and rename elsewhere, e.g. this function would become `on_learn_pkg`)?

Reply:
Agreed to the latter. I renamed `Pkg` to `LearnedPkg` and changed the method name to `on_learn_pkg`. However, the former is not correct. The `Share` can actually come from learners. That's why we give them a `LearnPkg` containing a share in the first place :)